aspnet/fundamentals/dependency-injection.rst

```diff
@@ -1,10 +1,277 @@
+
+.. code-block:: c#
+
+	// using Microsoft.Dnx.Runtime.Infrastructure
+	var hostingEnv = CallContextServiceLocator.Locator.ServiceProvider
+		.GetService(typeof(IHostingEnvironment)) as IHostingEnvironment;
+	string environmentName = hostingEnv.EnvironmentName;
+
+.. warning:: Prefer the use of dependency injection over service location in your application code, since it provides looser coupling and better testability. Service location may be the best (or only) choice in some framework or test code scenarios, however.
+
+Designing Your Services For Dependency Injection
+------------------------------------------------
+
+You should design your services to use dependency injection to get their collaborators. This means avoiding the use of stateful static method calls (which result in a code smell known as `static cling <http://deviq.com/static-cling/>`_) and the direct instantiation of dependent classes within your services. It may help to remember the phrase, `New is Glue <http://ardalis.com/new-is-glue>`_, when choosing whether to instantiate a type or to request it via dependency injection. By following the `SOLID Principles of Object Oriented Design <http://deviq.com/solid/>`_, your classes will naturally tend to be small, well-named and well-factored, and easily tested.
+
+What if you find that your classes tend to have way too many dependencies being injected? This is generally a sign that your class is trying to do too much, and is probably violating SRP - the `Single Responsibility Principle <http://deviq.com/single-responsibility-principle/>`_. See if you can refactor the class by moving some of its responsibilities into a new class. Keep in mind that your ``Controller`` classes should be focused on UI concerns, so business rules and data access implementation details should be kept in classes appropriate to these separate concerns.
+
+With regard to data access specifically, you can easily inject Entity Framework ``DbContext`` types into your controllers, assuming you've configured EF in your ``Startup`` class. However, I recommend avoiding direct dependency on ``DbContext`` in your application. Instead, depend on an abstraction (like a Repository interface), and restrict knowledge of EF (or any other specific data access technology) to the implementation of this interface. This will reduce the coupling between your application and a particular data access strategy, and will make :doc:`testing` your application code much easier.
```

Comment:
I don't really feel like this is out of place here. It could be made a little more generic, by using less terminology, but I'll leave that up to Steve. Feedback that I've gotten from customer visits that I've been on is that our users want guidance on leveraging DI.